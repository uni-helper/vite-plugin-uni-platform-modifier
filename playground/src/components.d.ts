/* eslint-disable */
/* prettier-ignore */
// @ts-nocheck
// Generated by vite-plugin-uni-components
// Read more: https://github.com/vuejs/core/pull/3399
import '@vue/runtime-core'

export {}

declare module '@vue/runtime-core' {
  export interface GlobalComponents {
    Counter: typeof import('./components/Counter.vue')['default']
    Footer: typeof import('./components/Footer.vue')['default']
    InputEntry: typeof import('./components/InputEntry.vue')['default']
    Logos: typeof import('./components/Logos.vue')['default']
  }
}
