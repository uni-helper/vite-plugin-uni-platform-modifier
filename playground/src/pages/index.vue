<script lang="ts" setup>
const handleDefaultClick = () => {
  console.log('default')
}
const handleH5Click = () => {
  console.log('h5')
}
</script>

<template>
  <div class="dark" class.mp-weixin="wx">
    <button v-text.h5="'hello'" v-text.h5.mp-weixin="'h5'" @click.h5="handleH5Click" class.h5="h5-class"
      class="default-class" @click="handleDefaultClick" />
  </div>
</template>
