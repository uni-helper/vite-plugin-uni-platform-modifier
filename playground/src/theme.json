{
  "light": {
    "bgColor": "#fcfcfc",
    "bgColorBottom": "#fcfcfc",
    "bgColorTop": "#ff6b00",
    "bgTxtStyle": "dark",
    "navBgColor": "#ff6b00",
    "navTxtStyle": "white",
    "tabFontColor": "#1f2937",
    "tabSelectedColor": "#ff6b00",
    "tabBgColor": "#fcfcfc",
    "tabBorderStyle": "black"
  },
  "dark": {
    "bgColor": "#181818",
    "bgColorBottom": "#181818",
    "bgColorTop": "#ff6b00",
    "bgTxtStyle": "light",
    "navBgColor": "#ff6b00",
    "navTxtStyle": "white",
    "tabFontColor": "#f3f4f6",
    "tabSelectedColor": "#ff6b00",
    "tabBgColor": "#181818",
    "tabBorderStyle": "white"
  }
}
