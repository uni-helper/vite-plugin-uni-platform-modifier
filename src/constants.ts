export const pluginName = 'vite-plugin-uni-platform-modifier'
