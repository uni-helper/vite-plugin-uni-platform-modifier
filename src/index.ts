import { type Plugin, createFilter, createLogger } from 'vite'
import type { ParentNode, TemplateChildNode } from '@vue/compiler-core'
import { builtInPlatforms, platform } from '@uni-helper/uni-env'

import { parse } from '@vue/compiler-dom'
import MagicString from 'magic-string'

function hasPlatformModifier(modifiers: string[]) {
  return modifiers.some(m => builtInPlatforms.includes(m as any))
}

const pluginName = 'vite-plugin-uni-platform-modifier'

export function VitePluginUniPlatformVOnModifier(): Plugin {
  const logger = createLogger(undefined, {
    prefix: pluginName,
  })
  return {
    name: pluginName,
    enforce: 'pre',
    transform(code, id) {
      const filter = createFilter('**/*.?(nu)vue')

      if (!filter(id))
        return

      const ast = parse(code)
      const rootTemplate = ast.children.find(node => node.type === 1 && node.tag === 'template') as ParentNode
      if (!rootTemplate)
        return
      const ms = new MagicString(code)
      function replaceProps(node: ParentNode | TemplateChildNode) {
        if (node.type !== 1)
          return

        const props = node.props.map((prop) => {
          const result = {
            ...prop.loc,
            name: '',
            modifiers: [] as string[],
          }
          if (prop.type === 6) {
            const [originalName, ...modifiers] = prop.name.split('.')
            result.name = originalName
            result.modifiers = modifiers
          }
          else {
            result.name = `v-${prop.name}${prop.arg?.loc.source ?? ''}`
            result.modifiers = prop.modifiers
          }
          return result
        })
        const groups = props.reduce((acc, cur) => {
          acc[cur.name] = acc[cur.name] ? [...acc[cur.name], cur] : [cur]
          return acc
        }, {} as Record<string, typeof props[0][]>)

        Object.keys(groups).forEach((name) => {
          const group = groups[name]
          const defaultProp = group.find(v => v.modifiers.length === 0 || !hasPlatformModifier(v.modifiers))
          const platformProps = group.filter(v => hasPlatformModifier(v.modifiers) && v.modifiers.includes(platform))
          const platformProp = platformProps[0]
          platformProps.slice(1).forEach((v) => {
            logger.warn(`${id}:${v.start.line}:${v.start.column} 属性 \`${name}\` 匹配到多个的平台修饰符，已忽略 ${v.source}`, {
              timestamp: true,
            })
          })
          const resultSource = platformProp?.source.replace(new RegExp(`\\.${builtInPlatforms.join('|\\.')}`, 'gm'), '') ?? defaultProp?.source ?? ''
          group.forEach(v => ms.remove(v.start.offset, v.start.offset))

          if (defaultProp)
            ms.overwrite(defaultProp.start.offset, defaultProp.end.offset, resultSource)
          else if (platformProp)
            ms.overwrite(platformProp.start.offset, platformProp.end.offset, resultSource)
        })

        node.children.forEach((n) => {
          replaceProps(n)
        })
      }
      replaceProps(rootTemplate)

      return {
        code: ms.toString(),
        map: ms.generateMap({
          file: id,
        }),
      }
    },
  }
}

export default VitePluginUniPlatformVOnModifier
