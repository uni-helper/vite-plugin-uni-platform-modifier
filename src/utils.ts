import { builtInPlatforms } from '@uni-helper/uni-env'
import type { AttributeNode, DirectiveNode, ParentNode, RootNode } from '@vue/compiler-core'

export function hasPlatformModifier(modifiers: string[]) {
  return modifiers.some(m => builtInPlatforms.includes(m as any))
}

export function normalizeVueProps(props: (AttributeNode | DirectiveNode)[]) {
  return props.map((prop) => {
    const result = {
      ...prop.loc,
      name: '',
      modifiers: [] as string[],
    }
    if (prop.type === 6) {
      const [originalName, ...modifiers] = prop.name.split('.')
      result.name = originalName
      result.modifiers = modifiers
    }
    else {
      result.name = `v-${prop.name}${prop.arg?.loc.source ?? ''}`
      result.modifiers = prop.modifiers
    }
    return result
  })
}

export function getRootTemplate(ast: RootNode) {
  return ast.children.find(node => node.type === 1 && node.tag === 'template') as ParentNode
}
