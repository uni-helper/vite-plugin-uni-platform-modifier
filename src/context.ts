import type { Logger } from 'vite'
import { createFilter, createLogger } from 'vite'

import type { ParentNode, TemplateChildNode } from '@vue/compiler-core'
import { builtInPlatforms, platform } from '@uni-helper/uni-env'

import { parse } from '@vue/compiler-dom'
import MagicString from 'magic-string'
import { pluginName } from './constants'
import { getRootTemplate, hasPlatformModifier, normalizeVueProps } from './utils'

export class Context {
  logger: Logger
  filter: (id: unknown) => boolean
  constructor() {
    this.logger = createLogger(undefined, {
      prefix: pluginName,
    })

    this.filter = createFilter('**/*.?(nu)vue')
  }

  transform(code: string, id: string) {
    const ast = parse(code)
    const rootTemplate = getRootTemplate(ast)
    if (!rootTemplate)
      return
    const ms = new MagicString(code)
    const replaceProps = (node: ParentNode | TemplateChildNode) => {
      if (node.type !== 1)
        return

      const props = normalizeVueProps(node.props)
      const groups = props.reduce((acc, cur) => {
        acc[cur.name] = acc[cur.name] ? [...acc[cur.name], cur] : [cur]
        return acc
      }, {} as Record<string, typeof props[0][]>)

      Object.keys(groups).forEach((name) => {
        const group = groups[name]
        const defaultProp = group.find(v => v.modifiers.length === 0 || !hasPlatformModifier(v.modifiers))
        const platformProps = group.filter(v => hasPlatformModifier(v.modifiers) && v.modifiers.includes(platform))
        const platformProp = platformProps[0]
        platformProps.slice(1).forEach((v) => {
          this.logger.warn(`${id}:${v.start.line}:${v.start.column} 属性 \`${name}\` 匹配到多个的平台修饰符，已忽略 ${v.source}`, {
            timestamp: true,
          })
        })
        const resultSource = platformProp?.source.replace(new RegExp(`\\.${builtInPlatforms.join('|\\.')}`, 'gm'), '') ?? defaultProp?.source ?? ''
        group.forEach(v => ms.remove(v.start.offset, v.start.offset))

        if (defaultProp)
          ms.overwrite(defaultProp.start.offset, defaultProp.end.offset, resultSource)
        else if (platformProp)
          ms.overwrite(platformProp.start.offset, platformProp.end.offset, resultSource)
      })

      node.children.forEach((n) => {
        replaceProps(n)
      })
    }
    replaceProps(rootTemplate)

    return {
      code: ms.toString(),
      map: ms.generateMap({
        file: id,
      }),
    }
  }
}
