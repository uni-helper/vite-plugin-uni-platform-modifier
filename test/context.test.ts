import { readFileSync } from 'node:fs'
import { relative, resolve } from 'node:path'
import { describe, expect, it, vi } from 'vitest'
import fg from 'fast-glob'
import { Context } from '../src/context'

vi.mock('@uni-helper/uni-env', () => {
  return {
    platform: 'mp-weixin',
    builtInPlatforms: ['h5', 'mp-weixin'],
  }
})

describe('context', async () => {
  const ctx = new Context()

  it('should return undefined if templateNode is not found', () => {
    const code = 'const x = 1;'
    const id = 'test.js'
    const result = ctx.transform(code, id)
    expect(result).toBeUndefined()
  })

  it('should return undefined if MagicString has not changed', () => {
    const code = '<template><div>{{ msg }}</div></template>'
    const id = 'test.vue'
    const result = ctx.transform(code, id)
    expect(result?.code).toBeUndefined()
  })

  const cwd = resolve(__dirname, './fixtures')
  const files = await fg('*.vue', {
    cwd,
    absolute: true,
  })

  // 循环测试每个文件
  for (const file of files) {
    it (`should transform ${relative(cwd, file)}`, () => {
      const code = readFileSync(file, 'utf-8')
      const id = file
      const result = ctx.transform(code, id)
      expect(result?.code).toMatchSnapshot()
    })
  }
})
