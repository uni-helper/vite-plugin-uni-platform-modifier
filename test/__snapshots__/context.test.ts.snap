// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`context > should transform fullAttr.vue 1`] = `
"<template>
  <div  is-link=\\"true\\" ></div>
</template>
"
`;

exports[`context > should transform noDefault.vue 1`] = `
"<template>
  <div  is-link=\\"true\\"></div>
</template>
"
`;
